lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -0,0 +1,754 @@
+                // a new overlay from the pending set is sufficient.
+                //
+                // Unwrapping is safe here because the set was already checked
+                // for emptiness.
+                let (kind, contents) = set.pending.pop_first().unwrap();
+                mapped.write_page(&contents);
+                set.active = kind;
+            }
+
+            Ok(())
+        } else {
+            // This overlay kind isn't active for this page. If it's pending,
+            // it's sufficient just to remove its entry from the pending map,
+            // since it has had no effect on guest memory.
+            //
+            // If there's no pending overlay of this kind, the caller goofed.
+            set.get_mut()
+                .pending
+                .remove(&kind)
+                .ok_or(OverlayError::NoOverlayOfKind(pfn.into(), kind))
+                .map(|_| ())
+        }
+    }
+
+    /// Moves the overlay of the supplied `kind` from `from_pfn` to `to_pfn`.
+    fn move_overlay(
+        &mut self,
+        from_pfn: Pfn,
+        to_pfn: Pfn,
+        kind: OverlayKind,
+        acc_mem: &MemAccessor,
+    ) -> Result<(), OverlayError> {
+        if from_pfn == to_pfn {
+            return Ok(());
+        }
+
+        let memctx =
+            acc_mem.access().ok_or(OverlayError::GuestMemoryInaccessible)?;
+
+        // Move the overlay by removing it from its old location and applying it
+        // to its new one. This will only work if an overlay of the requested
+        // kind does in fact exist at the source location, which means that it
+        // must have an active overlay set.
+        let Some(from_set) = self.overlays.get_mut(&from_pfn) else {
+            return Err(OverlayError::NoOverlaysActive(from_pfn.into()));
+        };
+
+        // The overlay contents may be stored in the table itself (for pending
+        // overlays) or in guest memory (for active overlays). Before bothering
+        // to obtain them, create a mapping for the target page to verify that
+        // the target PFN is valid. This mapping will be used later if this is
+        // the first overlay for the target PFN.
+        let mut to_mapping = to_pfn.map(&memctx)?;
+
+        // Get the page contents to transfer, but don't actually change the
+        // tables yet, since applying the new overlay might still fail if there
+        // is an existing overlay of this kind at the target PFN.
+        //
+        // It's possible to do this check now, modify the tables in-place, and
+        // then assert that adding the new overlay succeeded, but this sequence
+        // allows this function to reuse the existing overlay removal code. This
+        // is unlikely to be a bottleneck in practice because guests change
+        // their overlay locations so infrequently (and any guest that spends
+        // all its time moving overlays around has, in some ways, asked for
+        // it...).
```

Comment:
Reasonable!

Reply:
I still hope the guest has better things to do than spam its overlay MSRs, but this comment is gone with f09684e, along with cloning pending overlay contents :)